sql: maintain backref during restore of trigger

spilchen opened the pull request:
When doing a restore, we have code that removes triggers if there are any missing dependencies. Every reference has a forward reference and a backward reference. Removal of the triggers was only accounting for removal of the forward reference. This left dangling backreferences whenever trigger removal was done during restore.

This change corrects that by properly cleaning up any back reference that existed because of triggers.

It handles the following cases:
- if the removal of the trigger removes the last forward reference to a user defined type, update the types backref
- consider removal of any relation backref if the trigger had a forward ref to a relation. Generally, we remove all backrefs except in cases where the relation was for a sequence

Epic: none
Fixes: #147155

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/147306)
<!-- Reviewable:end -->

DrewKimball reviewed (commented):
> Reviewed 5 of 5 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147306)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @kev-cao and @mgartner)

___
*[`pkg/sql/catalog/rewrite/rewrite.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147306#-ORIbr_QAOLzB-lrOTmx:-ORIbr_QAOLzB-lrOTmy:bhg5fm2) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cc26c28d6bc3e20e9986caddbf617b7ec5559f/pkg/sql/catalog/rewrite/rewrite.go#L54)):*
> ```Go
> 	// because of trigger-related cleanup. This map is returned to the caller for
> 	// handling in a later pass.
> 	typeBackrefsToRemove = make(map[descpb.ID]map[descpb.ID]struct{})
> ```

Do we need something similar for function backrefs?
___
*[`pkg/sql/catalog/rewrite/rewrite.go` line 1192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147306#-ORIfL8u4mw1-n_ZqNb1:-ORIfL8u4mw1-n_ZqNb2:b1pgeey) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cc26c28d6bc3e20e9986caddbf617b7ec5559f/pkg/sql/catalog/rewrite/rewrite.go#L1192)):*
> ```Go
> 		trigger := &table.Triggers[i]
> 		if !foundAllDeps(trigger.DependsOn) || !foundAllDeps(trigger.DependsOnTypes) ||
> 			!foundAllDeps(trigger.DependsOnRoutines) {
> ```

Similar to above, why don't we have to handle routine back-references here?
___
*[`pkg/backup/restore_job.go` line 1160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147306#-ORIb6fJ-wvzaFnFLzhR:-ORIb6fJ-wvzaFnFLzhS:bd4gv6g) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cc26c28d6bc3e20e9986caddbf617b7ec5559f/pkg/backup/restore_job.go#L1160)):*
> ```Go
> 	}
> 
> 	// TODO(chengxiong): for now, we know that functions are not referenced by any
> ```

Is this comment stale?


<!-- Sent from Reviewable.io -->

spilchen committed:
sql: maintain backref during restore of trigger

When doing a restore, we have code that removes triggers if there are
any missing dependencies. Every reference has a forward reference and a
backward reference. Removal of the triggers was only accounting for
removal of the forward reference. This left dangling backreferences
whenever trigger removal was done during restore.

This change corrects that by properly cleaning up any back reference
that existed because of triggers.

It handles the following cases:
- if the removal of the trigger removes the last forward reference to a
user defined type, update the types backref
- consider removal of any relation backref if the trigger had a forward
ref to a relation. Generally, we remove all backrefs except in cases
where the relation was for a sequence

Epic: none
Fixes: #147155

spilchen reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147306)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, and @mgartner)

___
*[`pkg/backup/restore_job.go` line 1160 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147306#-ORIb6fJ-wvzaFnFLzhR:-ORLz7-T4J-qV8xmhARO:b-sigjli) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cc26c28d6bc3e20e9986caddbf617b7ec5559f/pkg/backup/restore_job.go#L1160)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Is this comment stale?
</blockquote></details>

Yes, parts of the comment were stale. Function descriptors are still only restored during full database restores, and they’re not referenced by other object types in a way that requires remapping. So collision handling for functions isn’t needed today.

For table-level restores, we rely on existing behavior: if the function already exists, the table restore requires dropping the dependent table first; if the function is missing, the restore fails unless `skip_missing_udfs` is set. I'll update the comment to reflect this.
___
*[`pkg/sql/catalog/rewrite/rewrite.go` line 54 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147306#-ORIbr_QAOLzB-lrOTmx:-ORLzQgaEna-JovHRvP_:bpfmhmb) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cc26c28d6bc3e20e9986caddbf617b7ec5559f/pkg/sql/catalog/rewrite/rewrite.go#L54)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Do we need something similar for function backrefs?
</blockquote></details>

Good question. I considered this before creating the PR. For function backrefs, I couldn’t find a scenario where we end up with a dangling reference.

In the case of restoring into an existing DB, if a table depends on a UDF, that table must be dropped before restore. Dropping it removes the backref from the function, so nothing stale remains. If the DB is empty and the function is missing, restore fails unless `skip_missing_udfs` is used. If that is used the function is simply skipped and not restored, which avoids any dangling reference.

Unlike types or relations, there's no path where a table restore alone would implicitly reintroduce a function or need to update its backrefs. You can’t selectively restore functions, and triggers that reference UDFs are dropped on restore anyway.

It's probably worth an update to the code to clarify. I'll add something to dropTriggerMissingDeps.
___
*[`pkg/sql/catalog/rewrite/rewrite.go` line 1192 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/147306#-ORIfL8u4mw1-n_ZqNb1:-ORM0Chj5ZF6cuX8JbOW:b-otilcs) ([raw file](https://github.com/cockroachdb/cockroach/blob/85cc26c28d6bc3e20e9986caddbf617b7ec5559f/pkg/sql/catalog/rewrite/rewrite.go#L1192)):*
<details><summary><i>Previously, DrewKimball (Drew Kimball) wrote…</i></summary><blockquote>

Similar to above, why don't we have to handle routine back-references here?
</blockquote></details>

See comment above.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
thanks for the helpful comments on this! the logic seems sound to me.

since this is dealing with RESTORE, please get a sign-off from the DR team. try posting in their slack channel and providing some background on the issue to solicit a review.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/147306#-:-ORO4AZs6ukoW_qLGQ25:b-7zcn3g)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @DrewKimball, @kev-cao, and @mgartner)



<!-- Sent from Reviewable.io -->

msbutler reviewed (approved):
Restore changes look fine!
